src/qibo/backends/abstract.py

```diff
@@ -81,291 +143,227 @@ def set_dtype(self, dtype: str):  # pragma: no cover
+            raise_error(
+                ValueError,
+                f"Unknown ``dtype`` ``{dtype}``. For this backend ({self}), "
+                + f"``dtype`` must be either one of the following string: {dtypes_str}, "
+                + f"or one of the following options: {self.numeric_types}",
+            )
 
-        Args:
-            device (str): Device such as '/CPU:0', '/GPU:0', etc.
-        """
-        raise_error(NotImplementedError)
+        if dtype != self.dtype:
+            self.dtype = dtype
 
-    @abc.abstractmethod
-    def set_threads(self, nthreads: int):  # pragma: no cover
-        """Set number of threads for CPU simulation.
+            if self.matrices:
+                self.matrices = self.matrices.__class__(self.dtype)
 
-        Args:
-            nthreads (int): Number of threads.
-        """
-        raise_error(NotImplementedError)
+    def set_seed(self, seed: Union[int, None]) -> None:
+        """Set the seed of the random number generator. Works in-place."""
+        self.engine.random.seed(seed)
 
-    @abc.abstractmethod
-    def cast(self, x, dtype=None, copy: bool = False):  # pragma: no cover
-        """Cast an object as the array type of the current backend.
+    def set_threads(self, nthreads: int) -> None:  # pragma: no cover
+        """Set number of threads for CPU backend simulations that accept it. Works in-place.
 
         Args:
-            x: Object to cast to array.
-            dtype (str or type, optional): data type of ``x`` after casting.
-                Options are ``"complex128"``, ``"complex64"``, ``"float64"``,
-                or ``"float32"``. If ``None``, defaults to ``Backend.dtype``.
-                Defaults to ``None``.
-            copy (bool, optional): If ``True`` a copy of the object is created in memory.
-                Defaults to ``False``.
+            nthreads (int): Number of threads.
         """
         raise_error(NotImplementedError)
 
-    @abc.abstractmethod
-    def is_sparse(self, x):  # pragma: no cover
-        """Determine if a given array is a sparse tensor."""
-        raise_error(NotImplementedError)
-
-    @abc.abstractmethod
-    def to_numpy(self, x):  # pragma: no cover
+    def to_numpy(self, array) -> "ndarray":  # pragma: no cover
         """Cast a given array to numpy."""
         raise_error(NotImplementedError)
 
-    @abc.abstractmethod
-    def compile(self, func):  # pragma: no cover
-        """Compile the given method.
+    ########################################################################################
+    ######## Methods related to data types                                          ########
+    ########################################################################################
 
-        Available only for the tensorflow backend.
-        """
-        raise_error(NotImplementedError)
+    @property
+    def complex64(self):
+        return getattr(self.engine, "complex64")
```

Comment:
Here and below: why `getattr(self.engine, "...")`?

If you have a constant string, `getattr(x, "s")` should be fully equivalent to `x.s`. Isn't it?